storage/remote/wal_watcher.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"regexp"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/tsdb"
+	"github.com/prometheus/tsdb/wal"
+	fsnotify "gopkg.in/fsnotify/fsnotify.v1"
+)
+
+// WALWatcher watches the TSDB WAL for a given StorageClient (remote write config).
+type WALWatcher struct {
+	storage *Storage
+	logger  log.Logger
+
+	currentSegment  *wal.Segment
+	walDir          string
+	lastForwardedTs prometheus.Gauge
+
+	dirChan        chan string
+	ctx            context.Context
+	cancel         context.CancelFunc
+	quit           chan struct{}
+	exitForwarding chan struct{}
+}
+
+// NewWALWatcher creates a new WAL watcher for a given remote write client.
+func NewWALWatcher(logger log.Logger, storage *Storage, dirChan chan string) *WALWatcher {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	w := &WALWatcher{
+		logger:         logger,
+		storage:        storage,
+		dirChan:        dirChan,
+		ctx:            ctx,
+		cancel:         cancel,
+		quit:           make(chan struct{}),
+		exitForwarding: make(chan struct{}),
+	}
+	w.lastForwardedTs = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "prometheus_remote_write_last_forwarded_timestamp",
+		Help: "Timestamp of the last time the WAL watcher sent a sample to queue managers.",
+	})
+	prometheus.MustRegister(w.lastForwardedTs)
+	return w
+}
+
+func (w *WALWatcher) setWALDir(dir string) {
+	w.logger.Log("msg", fmt.Sprintf("setting wal dir to %s", dir))
+	w.walDir = dir
+}
+
+func (w *WALWatcher) Start() {
+	go w.watch()
+}
+
+func (w *WALWatcher) Stop() {
+	// It seems that sometimes the quit channel is already closed.
+	defer func() {
+		if r := recover(); r != nil {
+			level.Error(log.With(w.logger)).Log("err", r)
+		}
+	}()
```

Comment:
This can't be committed.

Reply:
this will require a PR for tsdb, I'll work on that